src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -1996,55 +2003,107 @@ static private void FlattenDeconstructVariables(ImmutableArray<DeconstructionVar
+                {
+                    var variable = new OutDeconstructVarPendingInference(assignmentSyntax);
+                    analyzedArguments.Arguments.Add(variable);
+                    analyzedArguments.RefKinds.Add(RefKind.Out);
+                    outVars.Add(variable);
+                }
 
-            MethodSymbol deconstructMethod = (MethodSymbol)deconstructMember;
-            if (deconstructMethod.MethodKind != MethodKind.Ordinary)
-            {
-                Error(diagnostics, ErrorCode.ERR_MissingDeconstruct, boundRHS.Syntax, boundRHS.Type);
-                return null;
-            }
+                string methodName = "Deconstruct";
+                var memberAccess = BindInstanceMemberAccess(
+                                        receiverSyntax, receiverSyntax, receiver, methodName, rightArity: 0,
+                                        typeArgumentsSyntax: default(SeparatedSyntaxList<TypeSyntax>), typeArguments: default(ImmutableArray<TypeSymbol>),
+                                        invoked: true, diagnostics: diagnostics);
 
-            if (deconstructMethod.ParameterCount != numCheckedVariables)
-            {
-                Error(diagnostics, ErrorCode.ERR_DeconstructWrongParams, boundRHS.Syntax, deconstructMethod, numCheckedVariables);
-                return null;
-            }
+                memberAccess = CheckValue(memberAccess, BindValueKind.RValueOrMethodGroup, diagnostics);
+                memberAccess.WasCompilerGenerated = true;
+
+                if (memberAccess.Kind != BoundKind.MethodGroup)
+                {
+                    return MissingDeconstruct(receiver, assignmentSyntax, numCheckedVariables, diagnostics, out outPlaceholders, receiver);
+                }
 
-            if (deconstructMethod.Parameters.Any(p => p.RefKind != RefKind.Out))
+                // After the overload resolution completes, the last step is to coerce the arguments with inferred types.
+                // That step returns placeholder (of correct type) instead of the outVar nodes that were passed in as arguments.
+                // So the generated invocation expression will contain placeholders instead of those outVar nodes.
+                // Those placeholders are also recorded in the outVar for easy access below, by the `SetInferredType` call on the outVar nodes.
+                bag = DiagnosticBag.GetInstance();
+                BoundExpression result = BindMethodGroupInvocation(
+                                            receiverSyntax, receiverSyntax, methodName, (BoundMethodGroup)memberAccess, analyzedArguments, bag, queryClause: null,
+                                            allowUnexpandedForm: true);
+
+                result.WasCompilerGenerated = true;
+
+                if (bag.HasAnyErrors())
+                {
+                    return MissingDeconstruct(receiver, assignmentSyntax, numCheckedVariables, diagnostics, out outPlaceholders, result);
```

Comment:
I suspect that at this point the ```bag``` might contain use-site errors that are the actual source why binding failed. Not reporting them might actually make it harder to figure out why compilation fails. Consider to still include what we have in the ```bag``` to ```diagnostics```. Other scenarios could benefit from this too, neither overload is better, etc.